### PR TITLE
Allow cmd/zfs mount unmount of snapshots

### PR DESCRIPTION
**Pull request by lundman**
2 part.

Allow mount and unmount of snapshots using `zfs mount dataset@snapshot`. (and unmount).

Add `zfs_rollback_os()` call for post-rollback OS specific requirements.

### Motivation and Context

"zfs mount dataset@snapshot" as mounting of snapshot has to be done
manually from userland in macOS. There simply is no way to handle it
within the kernel.

In a future PR we will add events to `zed` to run snapshot scripts
that issue `zfs mount snapshot`, so that kernel can "trigger" mounting
events, and simulate ".zfs" on macOS.

In addition, as it is the only other change in this source file;

Add `zfs_rollback_os()` call to the rollback logic, so platforms can
do specific requirements. In our case, we need to kick Finder into
refreshing after rollback, or it will (continue to) display the old content.

Here we should make a decision;

As this PR will show, all new code are currently in `#ifdef __APPLE__` as to not impact upstream platforms.

But, in the case of `zfs_rollback_os()` perhaps it is acceptable to simply define empty functions for Linux and FreeBSD instead. We already do that in some places. Thoughts?

Would Linux and FreeBSD be also amenable to `zfs mount dataset@snapshot` ? It seems a perfectly sane use-case for users, but it does change the command-line semantics.

Discuss?

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: cmd/zfs/zfs_main.c [hunk 1]
```diff
@@ -7159,6 +7167,39 @@ share_mount(int op, int argc, char **argv)
 
 		(void) fclose(mnttab);
 	} else {
+#if defined(__APPLE__)
+		/*
+		 * OsX can not mount from kernel, users are expected to mount
```
[the remark this answers is not in the document]
Comment (lundman) [reply cmd/zfs/zfs_main.c:7172]: Oh we do, when kernel triggers the snapshot mount, the very tail end of that, is that a user-process calls `zfs mount`. 

Legacy versions of macOS version did no automatic snapshots, and it was up to the user to run `zfs mount`. It is for this reason we kept `zfs mount` as an option to `cmd/zfs`, as opposed to, loading half of `zfs` into (a newly invented) `cmd/zfs_mount` to trigger the same code that handles `zfs mount` with all the code-replication that entails.

So basically, we need a userland way to mount snapshots, and due to legacy macOS versions, keeping `zfs mount` might be less confusing to users(?)

**Review by pcd1193182 — APPROVED** [review]

**Review by behlendorf — COMMENTED** [review]
> in the case of zfs_rollback_os() perhaps it is acceptable to simply define empty functions for Linux and FreeBSD instead. We already do that in some places. Thoughts?

Yes, that's definitely preferred and generally it's what we've done whenever possible.  In fact, if possible (and not too ugly!) I'd suggest creating a `zfs_snapshot_unmount_os()` and any other stubs as needed to move this logic out of the common code.

**Comment by lundman** [comment]
By "enabling" mount/unmount commands for all platforms, although they are empty functions, at least two of the testers fail. The ones that check "bad arguments" return failure. It would be up to Linux and FreeBSD to decide if those commands should now work, or do nothing. I suppose I could change return 0 to return -1 or something?
